fuel-bft/src/reactor.rs

```diff
@@ -0,0 +1,690 @@
+use crate::{Consensus, HeightRound, Message, Metadata, Network, RoundValidator, State};
+
+use async_trait::async_trait;
+use fuel_crypto::PublicKey;
+
+use alloc::boxed::Box;
+use alloc::vec::Vec;
+use core::borrow::Borrow;
+use core::{cmp, fmt, task};
+
+/// State machine of the network.
+///
+/// Will consume messages from the peers, update the internal state and broadcast consensus, when
+/// applicable.
+///
+/// With [`Reactor::heartbeat`], will check if the current node is a round leader. If yes, will
+/// propose a block.
+#[async_trait]
+pub trait Reactor {
```

Comment:
I am considering making this a concrete struct. However, before that, there is a couple of things we need to figure out. But first, its better we give further clarification to some parts of the comment.

> For example `drain_queue` and `requeue` can be externally accessed in a breaking way, which wouldn't be necessary at all if Reactor was a concrete struct.

How can that be breaking? `requeue` is intended to insert a message that was lost. If we have some network fault (or whatever other reason), it can be easily inserted with this method without asking the reactor to instantly process the message, as in `receive_message`. What can be done to improve the clarity here is renaming `requeue` to `queue`. However, as in the BFT properties, there is no way the user can break the consensus by using this method.

I'm not sure either how can `drain_queue` be breaking. If the user opts to call `drain_queue` for some reason, the queue will be cleared. Maybe the consensus data was offline and the user wants to drop old accumulated messages before restarting the data backend? For what reason the user would call `drain_queue` without the intention of clearing the queue?

> In the case of `requeue` and `drain_queue`, while having these as trait methods allows them to be persisted somewhere like a database, that would be completely pointless since the `drain_queue` API forces them to all fit into an in-memory `Vec

The set of pending messages isn't intended to be big or persisted. In fact, it could be very volatile, unstable and unreliable. The properties of BFT allow us to simply drop messages at random.

However, there is no reason to abuse that property. The structure that will store the queued messages could be anything - a hashmap, some database, etc. The return type is a concrete implementation for convenience, but `Vec` can be created out of any other data structure.

One example is [HashMap::drain](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.drain). We can just `m.drain().collect()` and if the user opts to use a hashmap for whatever reason he has, then this method wouldn't change either. The enforced requirement here is that he returns a concrete structure of owned messages - because this is the intended behavior. He shouldn't create copies of the messages and keep them, he should just drain the set.

What we can do to improve the clarity is `fn drain_queue -> impl Iterator<Message>`.

Reply:
I agree it should be a concrete struct. I'm making an experiment here and will post/request a new round as soon as its ready